`portable` lazy allows `nonportable` thunk (#3436)
ocaml-flambda/flambda-backend: jane/doc/extensions/modes/reference.md

> # Lazy

Reviewer: ```suggestion
This document is a reference of how the modes interact with various other features of OCaml.

# Lazy
```

I just think it's nice to have a little intro.

`lazy e` contains a thunk that evaluates `e`, as well as a mutable cell to store the

> result of `e`. Upon construction, `lazy e` follows the mode of `e` and the comonadic mode

Reviewer: I'm not sure what "follows" means here. Do you mean that `lazy e`'s monadic modes are that of `e` and comonadic ones are those of the thunk? But what modes does the thunk have? I think that, in order to be useful, this documentation needs an example -- I don't understand it as written.

of the thunk. Upon destruction (forcing a lazy value), the result follows the mode of lazy
value. Currently, the thunk, the result and the `lazy` are all heap-allocated.

> Additionally, forcing a lazy value involves accessing the mutable cell and thus requires

Reviewer: This paragraph I understand. But it would probably also be helped by an example.

The goal of this document is to be a reasonably complete reference to the mode system in
OCaml.

<!-- CR zqian: For a gentler introduction, see [the introduction](intro.md). -->

The mode system in the compiler tracks various properties of values, so that certain
performance-enhancing operations can be performed safely. For example:
- Locality tracks escaping. See [the local allocations reference](../local/reference.md)
- Uniqueness and linearity tracks aliasing. See [the uniqueness reference](../uniqueness/reference.md)
- Portability and contention tracks inter-thread sharing.
    <!-- CR zqian: reference for portability and contention -->

# Lazy
`lazy e` contains a thunk that evaluates `e`, as well as a mutable cell to store the
result of `e`. Upon construction, the mode of `lazy e` is weaker than `e`. For example, if
`e` is `nonportable`, then `lazy e` cannot be `portable`. Upon destruction (forcing a lazy

> value), the result is weaker than the mode of lazy value. For example, forcing a

Reviewer: ```suggestion
value), the result is no stronger than the mode of lazy value. For example, forcing a
```

`nonportable` lazy value cannot give a `portable` result. Additionally, forcing a lazy
value involves accessing the mutable cell and thus requires the lazy value to be
`uncontended`.

> Currently, the above rules don't apply to the locality axis, because both the result and

Reviewer: This suggests, to me, that `force` can take a local value and return a global one. Is that true?

Author: yes that's true.

Reviewer: But that's not true in the stdlib. Should it be? This sentence suggests that it is.

Author: Hmm my impression is that `stdlib` doesn't contain any mode stuff? If we want we should put a better `force` in `stdlib_alpha`?
(But then, I think many stuff is missing in stdlib_alpha)

Reviewer: I think there is improvement to be found here, but we don't need to fix it now. Resolving.

the lazy value are heap-allocated, so they are always `global`.

> Additionally, upon construction, the comonadic fragment of `lazy e` is weaker than the

Reviewer: ```suggestion
Additionally, upon construction, the comonadic modes of `lazy e` are no stronger than the
```

thunk. The thunk is checked as `fun () -> e`, potentially closing over variables, which

> weakens its comonadic fragment. This rule doesn't apply to several axes:

Reviewer: ```suggestion
may affect its comonadic modes. This rule doesn't apply to several axes:
```

Author: Hmm, "weaken" is more precise than "affect", right?

Reviewer: Sure. But it looks like you've made this change.

> weakens its comonadic fragment. This rule doesn't apply to several axes:

Reviewer: It seems that this rule applies to exactly 0 of the comonadic axes. Maybe just remove and state more directly? That is, the expression `lazy e` is `global`, `portable`, `many`, and inherits its uniqueness and its contention from the `e` (if that's true).

Author: We just merged #3283 , and very soon #3098. Maybe they will interact with lazy in a special way as well, but it's certainly safe to have the current behavior.
